keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1458 @@
+2. The value of resources specified in the container’s `resources` stanza.
+
+In simple words, if `resources` is specified at pod-level, the values in pod-level `resources` stanza are given a preference in determining the QoS class, otherwise container level `resources` are considerted for QoS determination.
+
+* **Guaranteed QoS**
+
+    A pod is considered `Guaranteed`, if for each resource type, either of following is true:
+    * It is explicitly resourced (pod.spec.resources is set) for that resource type AND the pod's request is equal to its limits.
+    * It is implicitly resourced (pod.spec.resources is not set) AND all containers requests are equal to their limits.
+
+  **Note: If the pod spec includes `resources` stanza at pod level, but either requests or limits are missing for a resource, the pod will not be considered Guaranteed QoS, even if container-level requests and limits are present and are equal.**
+
+* **Burstable QoS Criteria**
+
+    A pod is considered `Burstable`, if for each resource type it doesn't meet the criteria for Guaranteed QoS for that resource, and either of following is true:
+
+    * It is explicitly resourced (pod.spec.resources is set) for that resource, AND request (less than or equal to limit) or limit is set.
+    * It is implicitly resourced for that resource AND for at least one container in the pod, request (less than or equal to limit) or limit is set. 
+  
+
+* **BestEffort QoS Criteria**
+
+    None of pod-level or container-level resources requests or limits are specified.
+
+
+#### OOM Killer Behavior
+With Pod level resources, the OOM Killer will evolve to consider the cumulative resource limits set for the entire Pod. If the aggregated
+memory usage of all the containers within a Pod exceeds the effective pod level limits, the entire Pod becomes a candidate for OOM Killing, even if individual containers are within their limits.
+
+Within a pod, the OOM Killer will maintain its existing behavior, considering container level limits (if set) and QoS classes, so a container
+exceeding its individual limits can still be terminated even if a pod as a whole is under its limit.
+
+#### Admission Controller
+
+* LimitRanger
+
+  LimitRanger today supports objects of type `Container`, `Pod` and `PersistentVolumeClaim`. It allows following constraints:
+  
+    * min (this option is to specify min usage constraint on the resource)
+    * max (this option is to specify max usage constraints on the resource)
+    * defaultRequest (this option is to specify default resource request for the named resource)
+    * default (this option is to specify default limit for the named resource)
+    * maxLimitRequestRatio (represents the max burst value of the named resource)
+  
+  default and defaultRequest are not supported by LimitRanger of type `Pod`. To incorporate Pod level resource specifications, LimitRanger of type `Pod` logic will be extended with following changes:
```

Comment:
what would be our long term recommendation in 5 years? Will we say - do defaulter and min/max for the Pod? Or keep it be on container level?

Reply:
will we ever need per-container defaulter in 5 years?

Reply:
I think users will likely have different preferences about setting pod vs container resources. Do you thin we need to be that prescriptive in this KEP?

Reply:
There is a consistency problem we need to think about. Mostly from the perspective on what people are trying to express with the limit ranger. E.g. if min set for container, but the total pod resources set to the value, should limit ranger decline the pod? Another consistency question: https://github.com/kubernetes/enhancements/pull/4678#discussion_r1645196655

Besides the consistency issues above, thinking on how people will be using limit ranger in future may help define it's API now.

Reply:
I'm still leaning towards not being overly prescriptive on these, or not making too many assumptions as long as there's a clear way for users to configure this as they want. 

@thockin for more input.

Reply:
I think we should let users use defaulters at both container and pod level and let them decide the right combination of values for defaults.